### PR TITLE
fix: Multi-output column expressions in frame `sort` method

### PR DESCRIPTION
**Pull request by alexander-beedie**
Several fixes relating to multi-output expressions (including selectors) in frame `sort` method:

* Better handle multi-output column expressions in `by` parameter; could raise an incorrect _"the amount of ordering booleans … does not match the number of series …"_ error depending on the number of such expressions passed and how many columns they actually selected.
* Don't `panic` if multi-output column expression evaluates to no columns - bail with a suitable error instead.
* Correctly align "nulls_last" & "descending" booleans when we expand multi-output expressions.

**Also:** standardised on "{:?}" format string for Rust-side `ColumnNotFound` error so we always get the quotes around the column name (which helps to see things like the empty string being used as a column name, which otherwise just looks like an incomplete error message). We had mixed usage here; sometimes it would show the quotes, sometimes it wouldn't (**update**: now added `polars_err!(col_not_found=name)` to centralise this).

## Examples

```python
from datetime import date
import polars.selectors as cs
import polars as pl

df = pl.DataFrame({
    "dts": [date(2077,10,3), date(2077,10,2), date(2077,10,2)],
    "strs": ["abc", "def", "ghi"],
    "vals": [10.5, 20.3, 15.7],
})
```

### Before

Panic if no matching cols:
```python
df.sort(pl.col("^x.*$"))
# PanicException: index out of bounds: the len is 0 but the index is 0
```
Failed to use selectors in `by`:
```python
df.sort(cs.temporal() | cs.binary(), cs.string(), cs.integer())
# ComputeError: the amount of ordering booleans: 3 does not match the number of series: 2
```

### After

Descriptive error if no matching cols:
```python
df.sort(pl.col("^x.*$"))
# ComputeError: No columns selected for sorting
```
Successful use of selectors in `by`:
```python
df.sort(cs.temporal() | cs.binary(), cs.string(), cs.integer())
# shape: (3, 3)
# ┌────────────┬──────┬──────┐
# │ dts        ┆ strs ┆ vals │
# │ ---        ┆ ---  ┆ ---  │
# │ date       ┆ str  ┆ f64  │
# ╞════════════╪══════╪══════╡
# │ 2077-10-02 ┆ def  ┆ 20.3 │
# │ 2077-10-02 ┆ ghi  ┆ 15.7 │
# │ 2077-10-03 ┆ abc  ┆ 10.5 │
# └────────────┴──────┴──────┘
```

Note that use of multi-output expressions in `by` may _or may not_ fail in the current build - it depends on how many such expressions there are in relation to how many columns actually get selected.

### CODE REVIEW

**Comment by ritchie46** [comment]
> Note that use of multi-output expressions in by may or may not fail in the current build - it depends on how many such expressions there are in relation to how many columns actually get selected.

That's wrong. The columns selected by the sorted should also be selected by the optimizer. Do you mean we don't do that atm?

File: crates/polars-core/src/frame/explode.rs [hunk 1]
```diff
@@ -308,7 +308,7 @@ impl DataFrame {
         let mut iter = on.iter().map(|v| {
             schema
                 .get(v)
-                .ok_or_else(|| polars_err!(ColumnNotFound: "{}", v))
+                .ok_or_else(|| polars_err!(ColumnNotFound: "{:?}", v))
```
Comment (ritchie46) [inline crates/polars-core/src/frame/explode.rs:311]: Can we add a `polars_err!(col_not_found = name)` shorthand in `polars_err`. Then the formatting is in one place.

File: py-polars/polars/lazyframe/frame.py [hunk 2]
```diff
@@ -1366,8 +1367,18 @@ def sort(
             )
 
         by = parse_into_list_of_expressions(by, *more_by)
-        descending = extend_bool(descending, len(by), "descending", "by")
-        nulls_last = extend_bool(nulls_last, len(by), "nulls_last", "by")
+        descending = extend_bool(descending, len(by), "descending", "by", condense=True)
+        nulls_last = extend_bool(nulls_last, len(by), "nulls_last", "by", condense=True)
+
+        if (len(descending) > 1 or len(nulls_last) > 1) and any(
```
Comment (ritchie46) [inline py-polars/polars/lazyframe/frame.py:1373]: We need to check this when we convert the DSL to the IR.

**Comment by alexander-beedie** [comment]
> That's wrong. The columns selected by the sorted should also be selected by the optimizer. Do you mean we don't do that atm?

It's because we currently check that the number of column expressions matches the vec lengths of nulls_last/descending _before_ those expressions are expanded (there's an incorrect assumption baked-in that one expression == one column).

Meant to park this in Draft while I track down the best place to handle this properly, as it's definitely not the Python check I added 😜

File: py-polars/polars/lazyframe/frame.py (hunk 2 above)
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py]: Done✌️

File: crates/polars-plan/src/plans/conversion/dsl_to_ir.rs [hunk 3]
```diff
@@ -393,10 +393,36 @@ pub fn to_alp_impl(
...
+                    .zip(sort_options.descending.iter().cycle()),
+            ) {
+                let exprs = expand_expressions(input, vec![c], lp_arena, expr_arena)
+                    .map_err(|e| e.context(failed_here!(sort)))?;
+
+                nulls_last.extend(std::iter::repeat(n).take(exprs.len()));
+                descending.extend(std::iter::repeat(d).take(exprs.len()));
+                expanded_cols.extend(exprs);
+            }
+            let sort_options = sort_options
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:418]: I was being a bit conservative there, on the (admittedly very unlikely) chance that the same object might get reused in some other context elsewhere 😆

File: crates/polars-plan/src/plans/conversion/dsl_to_ir.rs [hunk 4]
```diff
@@ -393,10 +393,36 @@ pub fn to_alp_impl(
...
+                .clone()
+                .with_nulls_last_multi(nulls_last.clone())
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:420]: Thx - left over from some earlier `dbg!` checks (and now dropped with `mut` sort_options) 👌

**Pull request merged**